From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hlib Hraif <hlib.hraif@gmail.com>
Date: Tue, 20 May 2026 12:00:00 +0000
Subject: [PATCH] crypto: tegra - Disable softirqs before finalizing request

Softirqs must be disabled when calling the finalization function on
a request. Additionally, move the finalize call before key invalidation
(which sleeps in host1x_syncpt_wait with BH enabled) to close a race
window where a softirq can enqueue new requests while cur_req is being
cleared, corrupting the crypto engine's request queue.

This manifests as:

  list_del corruption, ->next is NULL
  kernel BUG at lib/list_debug.c:49!
  crypto_dequeue_request -> crypto_pump_work [crypto_engine]
  15820000.crypto

Reported-by: Guangwu Zhang <guazhang@redhat.com>
Fixes: 0880bb3b00c8 ("crypto: tegra - Add Tegra Security Engine driver")
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Link: https://lore.kernel.org/all/aa_kPXrFeOmhydP6@gondor.apana.org.au/
[eberman: Port to nvidia-oot repo]
Signed-off-by: Elliot Berman <eberman@anduril.com>
[zengraf: Fix race between key invalidation and crypto finalization]
Signed-off-by: Hlib Hraif <hlib.hraif@gmail.com>

---

diff --git a/drivers/crypto/tegra/tegra-se-aes.c b/drivers/crypto/tegra/tegra-se-aes.c
index ff1c58c0..703de39c 100644
--- a/drivers/crypto/tegra/tegra-se-aes.c
+++ b/drivers/crypto/tegra/tegra-se-aes.c
@@ -6,6 +6,7 @@
 
 #include <nvidia/conftest.h>
 
+#include <linux/bottom_half.h>
 #include <linux/clk.h>
 #include <linux/dma-mapping.h>
 #include <linux/module.h>
@@ -279,7 +280,7 @@ static int tegra_aes_do_one_req(struct crypto_engine *engine, void *areq)
 	struct tegra_aes_ctx *ctx = crypto_skcipher_ctx(crypto_skcipher_reqtfm(req));
 	struct tegra_aes_reqctx *rctx = skcipher_request_ctx(req);
 	struct tegra_se *se = ctx->se;
-	unsigned int cmdlen, key1_id, key2_id;
+	unsigned int cmdlen, key1_id, key2_id, save_alg;
 	int ret;
 
 	/* Set buffer size as a multiple of AES_BLOCK_SIZE*/
@@ -338,13 +339,24 @@ out:
 	dma_free_coherent(ctx->se->dev, rctx->datbuf.size,
 			  rctx->datbuf.buf, rctx->datbuf.addr);
 
+	/*
+	 * Save values needed after finalization may free the request.
+	 * Finalize with BH disabled to prevent softirqs from enqueuing
+	 * new requests while cur_req is being cleared, which can corrupt
+	 * the crypto engine's request queue.
+	 */
+	save_alg = ctx->alg;
+
+	local_bh_disable();
+	crypto_finalize_skcipher_request(se->engine, req, ret);
+	local_bh_enable();
+
+	/* Key invalidation may sleep; safe now that finalization is done */
 	if (tegra_key_is_reserved(key1_id))
-		tegra_key_invalidate_reserved(ctx->se, key1_id, ctx->alg);
+		tegra_key_invalidate_reserved(se, key1_id, save_alg);
 
 	if (tegra_key_is_reserved(key2_id))
-		tegra_key_invalidate_reserved(ctx->se, key2_id, ctx->alg);
-
-	crypto_finalize_skcipher_request(se->engine, req, ret);
+		tegra_key_invalidate_reserved(se, key2_id, save_alg);
 
 	return 0;
 }
@@ -1219,6 +1231,7 @@ static int tegra_ccm_do_one_req(struct crypto_engine *engine, void *areq)
 	struct crypto_aead *tfm = crypto_aead_reqtfm(req);
 	struct tegra_aead_ctx *ctx = crypto_aead_ctx(tfm);
 	struct tegra_se *se = ctx->se;
+	unsigned int save_alg, save_key_id;
 	int ret;
 
 	rctx->src_sg = req->src;
@@ -1289,10 +1302,19 @@ outbuf_err:
 	dma_free_coherent(ctx->se->dev, rctx->inbuf.size,
 			  rctx->inbuf.buf, rctx->inbuf.addr);
 
-	if (tegra_key_is_reserved(rctx->key_id))
-		tegra_key_invalidate_reserved(ctx->se, rctx->key_id, ctx->alg);
+	/*
+	 * Save values needed after finalization may free the request.
+	 */
+	save_alg = ctx->alg;
+	save_key_id = rctx->key_id;
 
+	local_bh_disable();
 	crypto_finalize_aead_request(ctx->se->engine, req, ret);
+	local_bh_enable();
+
+	/* Key invalidation may sleep; safe now that finalization is done */
+	if (tegra_key_is_reserved(save_key_id))
+		tegra_key_invalidate_reserved(ctx->se, save_key_id, save_alg);
 
 	return 0;
 }
@@ -1303,6 +1325,7 @@ static int tegra_gcm_do_one_req(struct crypto_engine *engine, void *areq)
 	struct crypto_aead *tfm = crypto_aead_reqtfm(req);
 	struct tegra_aead_ctx *ctx = crypto_aead_ctx(tfm);
 	struct tegra_aead_reqctx *rctx = aead_request_ctx(req);
+	unsigned int save_alg, save_key_id;
 	int ret;
 
 	rctx->src_sg = req->src;
@@ -1372,10 +1395,19 @@ outbuf_err:
 	dma_free_coherent(ctx->se->dev, rctx->inbuf.size,
 			  rctx->inbuf.buf, rctx->inbuf.addr);
 
-	if (tegra_key_is_reserved(rctx->key_id))
-		tegra_key_invalidate_reserved(ctx->se, rctx->key_id, ctx->alg);
+	/*
+	 * Save values needed after finalization may free the request.
+	 */
+	save_alg = ctx->alg;
+	save_key_id = rctx->key_id;
 
+	local_bh_disable();
 	crypto_finalize_aead_request(ctx->se->engine, req, ret);
+	local_bh_enable();
+
+	/* Key invalidation may sleep; safe now that finalization is done */
+	if (tegra_key_is_reserved(save_key_id))
+		tegra_key_invalidate_reserved(ctx->se, save_key_id, save_alg);
 
 	return 0;
 }
@@ -1753,6 +1785,7 @@ static int tegra_cmac_do_one_req(struct crypto_engine *engine, void *areq)
 	struct crypto_ahash *tfm = crypto_ahash_reqtfm(req);
 	struct tegra_cmac_ctx *ctx = crypto_ahash_ctx(tfm);
 	struct tegra_se *se = ctx->se;
+	unsigned int save_alg, save_key_id;
 	int ret = -EINVAL;
 
 	if (!ctx->key_id) {
@@ -1772,10 +1805,19 @@ static int tegra_cmac_do_one_req(struct crypto_engine *engine, void *areq)
 		rctx->task &= ~SHA_FINAL;
 	}
 out:
-	if (tegra_key_is_reserved(rctx->key_id))
-		tegra_key_invalidate_reserved(ctx->se, rctx->key_id, ctx->alg);
+	/*
+	 * Save values needed after finalization may free the request.
+	 */
+	save_alg = ctx->alg;
+	save_key_id = rctx->key_id;
 
+	local_bh_disable();
 	crypto_finalize_hash_request(se->engine, req, ret);
+	local_bh_enable();
+
+	/* Key invalidation may sleep; safe now that finalization is done */
+	if (tegra_key_is_reserved(save_key_id))
+		tegra_key_invalidate_reserved(ctx->se, save_key_id, save_alg);
 
 	return 0;
 }
diff --git a/drivers/crypto/tegra/tegra-se-hash.c b/drivers/crypto/tegra/tegra-se-hash.c
index 71e3a72b..d1af00db 100644
--- a/drivers/crypto/tegra/tegra-se-hash.c
+++ b/drivers/crypto/tegra/tegra-se-hash.c
@@ -5,6 +5,7 @@
  */
 
 #include <nvidia/conftest.h>
+#include <linux/bottom_half.h>
 #include <linux/clk.h>
 #include <linux/dma-mapping.h>
 #include <linux/module.h>
@@ -447,7 +448,14 @@ static int tegra_sha_do_one_req(struct crypto_engine *engine, void *areq)
 		rctx->task &= ~SHA_FINAL;
 	}
 
+	/*
+	 * Finalize with BH disabled to prevent softirqs from enqueuing
+	 * new requests while cur_req is being cleared, which can corrupt
+	 * the crypto engine's request queue.
+	 */
+	local_bh_disable();
 	crypto_finalize_hash_request(se->engine, req, ret);
+	local_bh_enable();
 
 	return 0;
 }
